core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -342,17 +261,21 @@ class KafkaController(val config: KafkaConfig, zkUtils: ZkUtils, val brokerState
       partitionStateMachine.startup()
 
       // register the partition change listeners for all existing topics on failover
-      controllerContext.allTopics.foreach(topic => partitionStateMachine.registerPartitionChangeListener(topic))
+      controllerContext.allTopics.foreach(topic => registerPartitionModificationsListener(topic))
       info("Broker %d is ready to serve as the new controller with epoch %d".format(config.brokerId, epoch))
       maybeTriggerPartitionReassignment()
       maybeTriggerPreferredReplicaElection()
+      info("starting the controller scheduler")
+      kafkaScheduler.startup()
+      kafkaScheduler.schedule("controller-metric-task", () => controllerEventQueue.put(UpdateMetrics), period = 10, unit = TimeUnit.SECONDS)
+      if (config.deleteTopicEnable) {
+        kafkaScheduler.schedule("topic-deletion-progress-check-task", () => controllerEventQueue.put(TopicDeletionProgressCheck),
```

Comment:
Hmm, adding scheduled event into the queue just for checking seems over killing. Would it be better to just do this check in main event loop? This applies to other scheduled tasks like leader balancing. If we do this, we can get rid the scheduler thread completely.